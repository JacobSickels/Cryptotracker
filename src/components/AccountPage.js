import React from 'react';
import { connect } from 'react-redux';
import { Input } from 'react-materialize';
import { startEditCurrency } from '../actions/filters';
import sortExchanges from '../selectors/exchanges';

/*
    AccountPage is a React Component that contains the functionality on the Account Page

    This component contains a dropdown for updating the User's base currency
*/

export class AccountPage extends React.Component {

    constructor(props) {
        super(props);
        
        //Component state for updating the dropdown when it is clicked
        this.state = {
            currency: props.currency
        }

    }

    //This function is called when the dropdown is changed
    changeBaseCurrency = (e) => {
        //Finding currency object with same id from exchanges array
        const currency = this.props.exchanges.find((element) => {
            if(element.id === e.target.value){
                return element;
            }
        });
        //Updating component state to chosen currency
        this.setState({ currency });
        //Updating Redux state with chosen currency
        this.props.startEditCurrency(currency);
    }

    //Render function renders the components to the screen
    render() {
        return (
            <div className="page-container">
                <div className="account__title">
                    <div className="container">
                        <div className="account__title">
                            <h1>My Account Page</h1>
                        </div>
                    </div>
                </div>
                <div className="container">
                    <div className="account__option">
                        <h2>Set Base Currency</h2>
                        <Input s={5} type='select' value={this.state.currency.id} onChange={this.changeBaseCurrency}>
                            {
                                //Inline that function loads all the currency data into the dropdown as options
                                this.props.exchanges.map((currency) =>
                                    <option key={currency.id} value={currency.id}>{currency.name}</option>
                                )
                            }
                        </Input>
                    </div>
                </div>
            </div>
        );
    }
}

/*
    This function maps Redux state to props that are passed into the component
    
    This function passes sorted exchange data from Redux state, and the current
    base currency in the filters data from Redux state to the component.
*/
const mapStateToProps = (state, props) => {

    const sortedExchange = sortExchanges(state.exchange);
    return {
        exchanges: sortedExchange,
        currency: state.filters.currency
    };
};

/*
    This function takes Redux dispatch actions and maps them to a function
    This new function is passed as a property in the component and is called in changeBaseCurrency
*/
const mapDispatchToProps = (dispatch) => {
    return {
        startEditCurrency: (currency) => dispatch(startEditCurrency(currency))
    };
};

//This creates a Higher Order Component letting the component access the Redux state and dispatch actions
export default connect(mapStateToProps, mapDispatchToProps)(AccountPage);