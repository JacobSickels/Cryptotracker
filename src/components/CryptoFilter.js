import React from 'react';
import { connect } from 'react-redux';
import { DateRangePicker } from 'react-dates';
import moment from 'moment';
import { setStartDate, setEndDate } from '../actions/filters';

/*
    CryptoFilter is a React Component that contains a Date Range Picker
    This component is used for updating the startDate and endDate filters on Redux state
*/

export class CryptoFilter extends React.Component {

    //Component state for tracking if the DatePicker is focused
    state = {
        calendarFocused: null
    };

    //Calls Redux dispatch functions for chosen dates in DatePicker
    onDatesChange = ({ startDate, endDate }) => {
        
        //If startDate and endDate are null, the clear button was clicked
        //The current day is passed when the clear button is clicked
        if(!startDate && !endDate) {
            this.props.setStartDate(moment().startOf('day'));
            this.props.setEndDate(moment().endOf('day'));
        }
        else {
            this.props.setStartDate(startDate);
            this.props.setEndDate(endDate);
        }
    }

    //This is called when Date Picker focus is changed
    onFocusChange = (calendarFocused) => {
        this.setState(() => ({ calendarFocused }));
    }

    render() {
        return (
            <div className="date-container">
                <div className="container">
                    <DateRangePicker 
                        startDate={this.props.filters.startDate}
                        endDate={this.props.filters.endDate}
                        onDatesChange={this.onDatesChange}
                        focusedInput={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        minimumNights={0}
                        showClearDates={true}
                        isOutsideRange={(day) => {
                            const today = moment();
                            return day.isAfter(today) || day.isBefore("2018-03-01");
                        }}
                    />
                </div>
            </div>
        )
    }
}

/*

    This function maps Redux state to props that are passed into the component
    
    This function takes the Redux state.filters and maps it to a filters object
    This new object is passed as a property to the component

*/
const mapStateToProps = (state) => ({
    filters: state.filters
});

/*

    This function takes Redux dispatch actions and maps them to startSetDate and setEndDate
    These new functions are passed as properties and are called in onDatesChange

*/
const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate: (endDate) => dispatch(setEndDate(endDate))
});
//This creates a Higher Order Component letting the component access the Redux state and dispatch actions
export default connect(mapStateToProps, mapDispatchToProps)(CryptoFilter);