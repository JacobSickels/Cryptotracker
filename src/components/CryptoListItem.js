import React from 'react';
import { Link } from 'react-router-dom'; //Eventually linking to the specific cryptos page for more specific filtering
import moment from 'moment';
import momentTimezone from 'moment-timezone';


const CryptoListItem = (props) => (
    <div>
    {
        props.crypto.map(
            (entry) => <p key={entry.timestamp}> Amount: { entry.amount } Time: { moment(entry.timestamp).tz('America/Chicago').format() } </p>
        ) 
    }
    </div>
);
export default CryptoListItem;