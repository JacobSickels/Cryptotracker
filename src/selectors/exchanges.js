/*
    This is the sortExchanges method used in other Components
    This function is reponsible for sorting the exchange data alphabnetically
*/
export default (exchanges) => {

    //Copying exchanges array to temp array for manipulating
    const filteredExchanges = exchanges.slice(0);

    //Sorting filteredExchanges by name
    filteredExchanges.sort((a, b) => {
        if(a.name > b.name)
            return 1;
        else if( b.name > a.name)
            return -1;
        else   return 0;        
    });

    //Returning filteredExchanges array
    return filteredExchanges;
};

/*
    convertToExchange is responsible for converting an array of cryptocurrency
    data to the exchangeRate that is passed in.
*/
export const convertToExchange = (exchangeRate, cryptos) => {

    //Map function going through each object in cryptos array and converting it to the exchange rate
    const converted = cryptos.map((entry) => ({
        amount: (parseFloat(entry.amount) * exchangeRate).toFixed(2),
        timestamp: entry.timestamp
    }));

    //Returning converted cryptocurrency array
    return converted;
}