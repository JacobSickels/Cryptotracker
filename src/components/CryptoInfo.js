import React from 'react';
import { connect } from 'react-redux';
import moment from 'moment';
import { Row, Col } from 'react-materialize';
import { getMaxMinData } from '../selectors/cryptos';

/*
    CryptoInfo is a React Component that contains information on when it 
    was the best time to buy and sell during the filtering period.
*/

export class CryptoInfo extends React.Component {

    constructor(props) {
        super(props);
    }

    render() {
        return (
            <div className="chart-info">
                <div className="container">
                    <Row> 
                        <Col s={12} m={4} l={4}>
                            <h1> Max Increase </h1>
                            {this.props.maxDifference}
                        </Col>
                        <Col s={12} m={4} l={4}>
                            <h1>Time To Buy</h1> 
                            <em>{moment(this.props.minElement.timestamp).format('ha MMM Do YYYY')}</em>
                            <p>{this.props.minElement.amount}</p>
                        </Col>
                        <Col s={12} m={4} l={4}>
                            <h1>Time To Sell</h1>
                            <em>{moment(this.props.maxElement.timestamp).format('ha MMM Do YYYY')}</em>
                            <p>{this.props.maxElement.amount}</p>
                        </Col>
                    </Row>
                </div>
            </div>
        );
    }

}

/*

    This function maps Redux state to props that are passed into the component
    
    This function relies on the getMaxMinData function that resides in selectors/cryptos
    The data is seperated into a object and the key values from the object are passed as props

*/
const mapStateToProps = (state, props) => {
    const data = getMaxMinData(state.cryptos, state.filters, props.name.toLowerCase());
    return { ...data }
};
//This creates a Higher Order Component letting the component access the Redux state
export default connect(mapStateToProps)(CryptoInfo);