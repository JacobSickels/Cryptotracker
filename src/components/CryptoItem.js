import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import { connect } from 'react-redux';
import { ResponsiveContainer, LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip, Legend } from 'recharts';
import { selectCryptos } from '../selectors/cryptos';
import { convertToExchange } from '../selectors/exchanges';
import getTrendData from '../selectors/trend';

/*
    
    CryptoItem is a React Component that contains all of the functionality needed
    to graph the specific cryptocurrency's data as well as showing the current price
    and current percentage increase for the cryptocurrency

*/


export class CryptoItem extends React.Component {
    
    constructor(props) {
        super(props);

        //Component state to keep track of showing and hiding trend line.
        this.state = {
            showTrendLine: true
        }
    }

    render() {
        return (
            <div className="chart">
                <div className="chart-title">
                    <div className="container">
                        <div className="chart-title--container">
                            <div>
                                <Link className="chart-title__link" to={`/filter/${this.props.name.toLowerCase()}`} >
                                    <h1> {this.props.name} </h1>
                                </Link>
                            </div>
                            <div className="chart-title__amount">
                                <h1>
                                    {this.props.currency_symbol} {this.props.current.amount}
                                </h1>
                            </div>
                            <div className="chart-title--percentage">
                                 <h1 style={{
                                     color:
                                        (this.props.percentage >= 0) ? 'green' : 'red'
                                     }}>
                                    {this.props.percentage.toFixed(2)}% 
                                </h1>
                            </div>
                        </div>
                    </div>
                </div>
                <div className="container">
                    <div className="button-container">
                        <button className="button--trend"
                            onClick={() => {
                                //Function for updating state to show or hide trend line
                                const show = this.state.showTrendLine;
                                this.setState({showTrendLine: !show})
                            }}>
                        {
                            (this.state.showTrendLine) ? 'Hide Trend' : 'Show Trend'
                        }      
                        </button>
                    </div>
                    <div className="chart-container">
                        <ResponsiveContainer height='100%' width='100%'>
                            <LineChart data={this.props.data} margin={{top: 5, right: 10, left: 30, bottom: 5}}>
                                <XAxis dataKey="timestamp"/>
                                <YAxis tickSize={10} dataKey="amount" domain={["dataMin", "dataMax"]} allowDataOverflow={true} />
                                <CartesianGrid strokeDasharray="3 3" stroke="#FFDDED"/>
                                <Tooltip/>
                                <Legend />
                                {
                                    //Inline function for showing and hiding trend line on graph
                                    (this.state.showTrendLine) && (
                                        <Line dot={false} type="monotone" dataKey="trend" stroke="#E477D4" strokeWidth={4}/>
                                    )
                                }
                                <Line dot={false} type="monotone" dataKey="amount" stroke="#7F4276"/>
                            </LineChart>
                        </ResponsiveContainer>
                    </div>
                </div>
            </div>
        );
    };
}


/*

    This function maps Redux state to props that are passed into the component
    
    This function relies on the selectCryptos function that resides in selectors/cryptos
    Crypto data is selected in the filtering range for the specific crypto
    The selected data is converted to the currency defined by the exchange rate
    Then the data is passed as props to the component

*/
const mapStateToProps = (state, props) => {

    let crypto = selectCryptos(state.cryptos, state.filters)[props.name.toLowerCase()];

    const exchangeRate = parseFloat(state.filters.currency.exchange_rate);

    //Converting crypto amounts to the exchange rate chosen fron Redux state
    crypto = convertToExchange(exchangeRate, crypto);

    return {
        ...getTrendData(crypto),
        currency_symbol: state.filters.currency.symbol //The currency symbol for the base currency
    }
    
};

//This creates a Higher Order Component letting the component access the Redux state
export default connect(mapStateToProps)(CryptoItem);