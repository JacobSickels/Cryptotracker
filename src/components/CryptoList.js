import React from 'react';
import CryptoItem from './CryptoItem';

/*
    CryptoList is a wrapper React Component used for the Dashboard Page
    This component renders a CryptoItem component for each name that is passed
    to it. The Dashboard page renders Bitcoin, Litecoin, and Ethereum data in 
    a sequence and this is the component that achieves that functionality.    

    This component doesn't have the JavaScript ES6 class definition for a React Component
    because it doesn't require all of the other functionality components of that nature need.
*/

const CryptoList = () => (
        <div>
            <CryptoItem name='Bitcoin'/>
            <CryptoItem name='Litecoin'/> 
            <CryptoItem name='Ethereum'/>
        </div>
);

export default CryptoList;