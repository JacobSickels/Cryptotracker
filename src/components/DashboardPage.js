import React from 'react';
import { connect } from 'react-redux';
import CryptoList from './CryptoList';
import { setDefaultFilter } from '../actions/filters';

/*
    DashboardPage is a React Component that is responsible for resetting the Redux filters
    to the current day when it is rendered, as well as rendering an instance of CryptoList.
*/

class DashboardPage extends React.Component {

    constructor(props) {
        super(props);
    }

    //Called before the component renders
    componentWillMount() {
        //Dashboard should only show the current day of data. 
        //This dispatch resets the filters to the current day
        this.props.setDefaultFilter();
    }

    render() {
        return (
            <div>
                <div className="page-container">
                    <CryptoList />
                </div>
            </div>
        );
    }
}

/*

    This function takes Redux dispatch actions and maps them to a function
    This new function is passed as a property in the component and is called in componentWillMount

    setDefaultFilter Redux action resets the startDate and endDate filters to the current day
*/
const mapDispatchToProps = (dispatch) => ({
    setDefaultFilter: () => dispatch(setDefaultFilter())
});

//This creates a Higher Order Component letting the component access the Redux dispatch actions
export default connect(null, mapDispatchToProps)(DashboardPage);