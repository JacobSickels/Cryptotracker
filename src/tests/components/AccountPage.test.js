import React from 'react';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';

import { AccountPage } from '../../components/AccountPage';
import { Input } from 'react-materialize';
import exchanges from '../fixtures/exchanges';

let store;
let wrapper;
let mockStartEditCurrency;

beforeEach(() => {
    const createMockStore = configureMockStore([thunk]);
    mockStartEditCurrency = jest.fn();
    store = createMockStore();
});

test('should render account page correctly', () => {
    const wrapper = shallow(
        <AccountPage store={store} currency={exchanges[0]} startEditCurrency={mockStartEditCurrency} exchanges={exchanges}/>
    );
    expect(wrapper).toMatchSnapshot();
});

test('should handle onInputChange', () => {
    
    const currency = {
        "exchange_rate": "106.31",
        "id": "ALL",
        "name": "Albanian Lek",
        "symbol": "Lek"
    };

    const wrapper = shallow(
        <AccountPage store={store} currency={currency} startEditCurrency={mockStartEditCurrency} exchanges={exchanges}/>
    );

    wrapper.find('Input').simulate('change', {
        target: { value: currency.id }
    });
    expect(wrapper.state('currency')).toEqual(currency);
});
