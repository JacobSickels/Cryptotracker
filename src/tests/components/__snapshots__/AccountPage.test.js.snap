// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render account page correctly 1`] = `
<div
  className="page-container"
>
  <div
    className="account__title"
  >
    <div
      className="container"
    >
      <div
        className="account__title"
      >
        <h1>
          My Account Page
        </h1>
      </div>
    </div>
  </div>
  <div
    className="container"
  >
    <div
      className="account__option"
    >
      <h2>
        Set Base Currency
      </h2>
      <Input
        checked={false}
        onChange={[Function]}
        s={5}
        type="select"
        value="AED"
      >
        <option
          key="AED"
          value="AED"
        >
          United Arab Emirates Dirham
        </option>
        <option
          key="AFN"
          value="AFN"
        >
          Afghan Afghani
        </option>
        <option
          key="ALL"
          value="ALL"
        >
          Albanian Lek
        </option>
        <option
          key="AMD"
          value="AMD"
        >
          Armenian Dram
        </option>
        <option
          key="ANG"
          value="ANG"
        >
          Netherlands Antillean Gulden
        </option>
        <option
          key="AOA"
          value="AOA"
        >
          Angolan Kwanza
        </option>
        <option
          key="ARS"
          value="ARS"
        >
          Argentine Peso
        </option>
      </Input>
    </div>
  </div>
</div>
`;
